### DISCUSSION TITLE
Changing Prev and Next on cmp-nvim not working

### DISCUSSION THREAD

**Discussion by dalssaso** [NvChad/NvChad#2028] [Q&A] [answered]
Hi.
I'm used to use the `<C-j>` and `<C-k>` to go to previous and next options in auto complete and i've been trying to set it up but it doesn't seem to be working. (I'm might have missed something)

Here's my config

* Overrides

```lua
M.cmp = {
  options = {
    mapping = {
    ["<C-k>"] = require("cmp").mapping.select_prev_item(),
    ["<C-j>"] = require("cmp").mapping.select_next_item(),
    ["<Tab>"] = require("cmp").mapping(function(fallback)
      if require("cmp").visible() then
        require("cmp").select_next_item()
      else
        fallback()
      end
    end, {
      "i",
      "s",
    }),
    }
  }
}
```

* Plugins

```lua
  {
    "hrsh7th/nvim-cmp",
    opts = overrides.cmp
  },

```

* Mappings

```lua
M.disabled = {
  i = {
    ["<C-h>"] = "",
    ["<C-l>"] = "",
    ["<C-j>"] = "",
    ["<C-k>"] = "",
  },
}
```

As I said, i think I'm missing something but I don't really understand how check the current values that are being passed to the plugin itself

Thanks

**Comment by siduck** [comment, marked as answer]
no need to disable our c-jkl in insert mode. in Neovim we can have the same mappings for a buffer so the mappings would work in cmp for sure!

```lua
{
    "hrsh7th/nvim-cmp",
    opts = {
      mapping = {
        ["<C-k>"] = require("cmp").mapping.select_prev_item(),
        ["<C-j>"] = require("cmp").mapping.select_next_item(),
      },
    },
  }
  ```

**Reply by dalssaso** [reply]
:man_facepalming: i  added the `options` and that screwed me up

Thanks @siduck

**Reply by siduck** [reply]
@dalssaso however there's an issue with that, all cmp modules will load at the start ( check with --startuptime ) which isnt good 

```lua
 {
    "hrsh7th/nvim-cmp",
    opts = function()
      local default_config = require "plugins.configs.cmp"

      local custom_changes = {
        mapping = {
          ["<C-k>"] = require("cmp").mapping.select_prev_item(),
          ["<C-j>"] = require("cmp").mapping.select_next_item(),
        },
      }

      return vim.tbl_deep_extend("error", default_config, custom_changes)
    end,
  }
```

@folke probably knows a better solution :thinking:

**Reply by folke** [reply]
@siduck two options:

Keep opts as a table, and just wrap the functions in a closure:

```lua
{
  "hrsh7th/nvim-cmp",
  opts = {
    mapping = {
      ["<C-k>"] = function()
        return require("cmp").mapping.select_prev_item()
      end,
      ["<C-j>"] = function()
        require("cmp").mapping.select_next_item()
      end,
    },
  },
}
```

Do like your suggestion, but assuming this is overriding the default spec,
the opts of the original spec are passed as the second parameter, so
no need to require those defaults again.

```lua
{
  "hrsh7th/nvim-cmp",
  opts = function(_, opts)
    local custom_changes = {
      mapping = {
        ["<C-k>"] = require("cmp").mapping.select_prev_item(),
        ["<C-j>"] = require("cmp").mapping.select_next_item(),
      },
    }
    return vim.tbl_deep_extend("error", opts, custom_changes)
  end,
}
```

**Reply by siduck** [reply]
@folke thanks for the solution, btw only the 2nd one works

**Reply by folke** [reply]
Sorry, should probably be somrthing like this then:

```lua
{
  "hrsh7th/nvim-cmp",
  opts = {
    mapping = {
      ["<C-k>"] = function(...)
        return require("cmp").mapping.select_prev_item(...)
      end,
      ["<C-j>"] = function(...)
        require("cmp").mapping.select_next_item(...)
      end,
    },
  },
}
```

I forgot the `...`

**Reply by siduck** [reply]
> Sorry, should probably be somrthing like this then:
> 
> ```lua
> {
>   "hrsh7th/nvim-cmp",
>   opts = {
>     mapping = {
>       ["<C-k>"] = function(...)
>         return require("cmp").mapping.select_prev_item(...)
>       end,
>       ["<C-j>"] = function(...)
>         require("cmp").mapping.select_next_item(...)
>       end,
>     },
>   },
> }
> ```
> 
> I forgot the `...`

tried, didnt work! can you test it at your end?
